src/cpp/handler/handlerengine.cpp

```diff
@@ -2474,9 +2474,9 @@ class HandlerEngine::Private : public QObject
 			all.httpResponseMessagesSent += qMax(p.httpResponseMessagesSent, 0);
 		}
 
-		report = ControlRequest::report(proxyControlClient, all, this);
-		finishedConnection[report] = report->finished.connect(boost::bind(&Private::report_finished, this, boost::placeholders::_1));
-		deferreds += report;
+		report = std::unique_ptr<Deferred>(ControlRequest::report(proxyControlClient.get(), all));
+		finishedConnection[report.get()] = report->finished.connect(boost::bind(&Private::report_finished, this, boost::placeholders::_1));
+		deferreds[report.get()] = std::move(report);
```

Comment:
storing the pointer in two places like this doesn't really work with unique_ptr I don't think. I assume the `std::move` nulls the `report`. it looks like `report` is only used in `stats_reported` to guard against starting multiple report rpc calls, and it only checks if it's non-null, so maybe just get rid of `report` and replace it with `bool reporting`.

Reply:
`std::move` is to move ownership of that particular unique_ptr to deferreds.
we still need `report` to keep track of `finishedConnection` thats based on report.

Reply:
[unique_ptr's docs](https://en.cppreference.com/w/cpp/memory/unique_ptr/unique_ptr) for its move constructor:

```c++
unique_ptr( unique_ptr&& u ) noexcept;             (5)
```

> 5) Constructs a unique_ptr by transferring ownership from u to *this and stores the null pointer in u.

so the `if(report)` will always return false, and we'll need some other way to track the finished connection

Reply:
perhaps `shared_ptr` could be used here, since that's logically what the original code was to doing (two handles to the same data), although it would be nice to find a way to not need that

`finishedConnection` might be fine, since it's populated while `report` is still valid, and erase is called with a slot argument instead of `report`